In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("lab04-cs104.ipynb")

# Lab 4: Data Types and Arrays
Welcome to Lab 4.

So far, we've used Python to manipulate numbers and work with tables. But we need to discuss data types to deepen our understanding of how to work with data in Python.

In this lab, you'll first see how to represent and manipulate another fundamental type of data: text.  A piece of text is called a *string* in Python. You'll also see how to work with *arrays* of data. An array could contain all the numbers between 0 and 100 or all the words in the chapter of a book. Lastly, you'll create tables and practice analyzing them with your knowledge of table operations. 

[Python Reference](http://data8.org/sp22/python-reference.html) has information that will be useful for this lab.

Let's begin by setting up the imports and tests by running the cell below and the cell above.

Set up the tests and imports by running the cell below.

In [2]:
# Just run this cell
import numpy as np
import math
from datascience import *

---
## Lab Check-off

Explain `np.arange`. What are its inputs/arguments and what are its outputs? Feel free to reference the following code: `np.arange(0,10,2)`.


---

In [100]:

np.arange(0,10, 2)

# 1st argument is index, 2nd is max, 3rd is the increments it will take.
# uniformed spaced values



array([0, 2, 4, 6, 8])

# 1. Text
Programming doesn't just concern numbers. Text is one of the most common data types used in programs. 

Text is represented by a **string value** in Python. The word "string" is a programming term for a sequence of characters. A string might contain a single character, a word, a sentence, or a whole book.

To distinguish text data from actual code, we demarcate strings by putting quotation marks around them. Single quotes (`'`) and double quotes (`"`) are both valid, but the types of opening and closing quotation marks must match. The contents can be any sequence of characters, including numbers and symbols. 

We've seen strings before in `print` statements.  Below, two different strings are passed as arguments to the `print` function.

In [3]:
print("I <3", 'Data Science')

I <3 Data Science


Just as names can be given to numbers, names can be given to string values.  The names and strings aren't required to be similar in any way. Any name can be assigned to any string.
(Although this example is not recommended, since it is confusing!)

In [4]:
one = 'two'
plus = '*'
print(one, plus, one)

two * two


**Question 1.1.** Yuri Gagarin was the first person to travel through outer space.  When he emerged from his capsule upon landing on Earth, he [reportedly](https://en.wikiquote.org/wiki/Yuri_Gagarin) had the following conversation with a woman and girl who saw the landing:

    The woman asked: "Can it be that you have come from outer space?"
    Gagarin replied: "As a matter of fact, I have!"

The cell below contains unfinished code.  Fill in the `...`s so that it prints out this conversation *exactly* as it appears above.

<!--
BEGIN QUESTION
name: q11
-->

In [5]:
woman_asking = 'The woman asked:'
woman_quote = '"Can it be that you have come from outer space?"'
gagarin_reply = 'Gagarin replied:'
gagarin_quote = '"As a matter of fact, I have!"'

print(woman_asking, woman_quote)
print(gagarin_reply, gagarin_quote)

The woman asked: "Can it be that you have come from outer space?"
Gagarin replied: "As a matter of fact, I have!"


In [6]:
grader.check("q11")

q11 results: All test cases passed!

## 1.1. String Methods

Strings can be transformed using **methods**. Recall that methods and functions are not the same thing. Here is the textbook section on string methods: [4.2.1 String Methods](https://inferentialthinking.com/chapters/04/2/1/String_Methods.html?highlight=methods). 

Here's a sketch of how to call methods on a string:

    <expression that evaluates to a string>.<method name>(<argument>, <argument>, ...)
    
One example of a string method is `replace`, which replaces all instances of some part of the original string (or a *substring*) with a new string. 

    <original string>.replace(<old substring>, <new substring>)
    
`replace` returns (evaluates to) a new string, leaving the original string unchanged.
    
Try to predict the output of this example, then run the cell!

In [7]:
# Replace one letter
bag = 'bag'
print(bag.replace('g', 't'), bag)

bat bag


`replace` replaces all instances, not just the first.

In [8]:
'alabama'.replace('a','a!')

'a!la!ba!ma!'

And it can replace strings of any length.

In [9]:
s = 'The father of my father is my grandfather.'
s.replace('father','mother')

'The mother of my mother is my grandmother.'

You can call functions on the results of other functions.  For example, `max(abs(-5), abs(3))` evaluates to 5.  Similarly, you can call methods on the results of other method or function calls.

You may have already noticed one difference between functions and methods - a function like `max` does not require a `.` before it's called, but a string method like `replace` does. Here's a handy [Python Reference](http://data8.org/sp22/python-reference.html).  It's a good idea to refer to this whenever you're unsure of how to call a function or method.

In [10]:
# Calling replace on the output of another call to replace
'train'.replace('t', 'ing').replace('in', 'de')

'degrade'

Here's a picture of how Python evaluates a "chained" method call like that:

<img src="chaining_method_calls.png"/>

**Question 1.1.1.** Use `replace` to transform the string `'hitchhiker'` into `'matchmaker'`. Assign your result to `new_word`.

<!--
BEGIN QUESTION
name: q111
-->

In [11]:
new_word = 'matchmaker'.replace('hitch', 'hiker')
new_word

'matchmaker'

In [12]:
grader.check("q111")

q111 results: All test cases passed!

There are many more string methods in Python, but most programmers don't memorize their names or how to use them.  In the "real world," people usually just search the internet for documentation and examples. A complete [list of string methods](https://docs.python.org/3/library/stdtypes.html#string-methods) appears in the Python language documentation. Developing your own primers of important commands, like our Python reference, is also recommended.

## 1.2. Converting to and from Strings

Strings and numbers are different *types* of values, even when a string contains the digits of a number. For example, evaluating the following cell causes an error because an integer cannot be added to a string.

In [13]:
8 + "8"

TypeError: unsupported operand type(s) for +: 'int' and 'str'

However, there are built-in functions to convert numbers to strings and strings to numbers. Some of these built-in functions have restrictions on the type of argument they take:

|Function |Description|
|-|-|
|`int`|Converts a string of digits or a float to an integer ("int") value|
|`float`|Converts a string of digits (perhaps with a decimal point) or an int to a decimal ("float") value|
|`str`|Converts any value to a string|

Try to predict what data type and value `example` evaluates to, then run the cell.

In [14]:
example = 8 + int("10") + float("8")

print(example)
print("This example returned a " + str(type(example)) + "!")

26.0
This example returned a <class 'float'>!


Suppose you're writing a program that looks for dates in a text, and you want your program to find the amount of time that elapsed between two years it has identified.  It doesn't make sense to subtract two texts, but you can first convert the text containing the years into numbers.

**Question 1.2.1.** Finish the code below to compute the number of years that elapsed between `one_year` and `another_year`.  Don't just write the numbers `1618` and `1648` (or `30`); use a conversion function to turn the given text data into numbers.

<!--
BEGIN QUESTION
name: q121
-->

In [15]:
# Some text data:
one_year = "1618"
another_year = "1648"

# Complete the next line.  Note that we can't just write:
#   another_year - one_year
# If you don't see why, try seeing what happens when you
# write that here.
difference = int(another_year) - int(one_year)
difference

30

In [16]:
grader.check("q121")

q121 results: All test cases passed!

## 1.3. Passing strings to functions

String values, like numbers, can be arguments to functions and can be returned by functions. 

The function `len` (derived from the word "length") takes a single string as its argument and returns the number of characters (including spaces) in the string.

Note that it doesn't count *words*. `len("one small step for man")` evaluates to 22 characters, not 5 words.

**Question 1.3.1.**  Use `len` to find the number of characters in the long string in the next cell.  Characters include things like spaces and punctuation. Assign `sentence_length` to that number.

(The string is the first sentence of the English translation of the French [Declaration of the Rights of Man](http://avalon.law.yale.edu/18th_century/rightsof.asp).)  

<!--
BEGIN QUESTION
name: q131
-->

In [17]:
a_very_long_sentence = "The representatives of the French people, organized as a National Assembly, believing that the ignorance, neglect, or contempt of the rights of man are the sole cause of public calamities and of the corruption of governments, have determined to set forth in a solemn declaration the natural, unalienable, and sacred rights of man, in order that this declaration, being constantly before all the members of the Social body, shall remind them continually of their rights and duties; in order that the acts of the legislative power, as well as those of the executive power, may be compared at any moment with the objects and purposes of all political institutions and may thus be more respected, and, lastly, in order that the grievances of the citizens, based hereafter upon simple and incontestable principles, shall tend to the maintenance of the constitution and redound to the happiness of all."
sentence_length = len(a_very_long_sentence)
sentence_length

896

In [18]:
grader.check("q131")

q131 results: All test cases passed!

# 2. Arrays

Computers are most useful when you can use a small amount of code to *do the same action* to *many different things*.

For example, in the time it takes you to calculate the 18% tip on a restaurant bill, a laptop can calculate 18% tips for every restaurant bill paid by every human on Earth that day.  (That's if you're pretty fast at doing arithmetic in your head!)

**Arrays** are how we put many values in one place so that we can operate on them as a group. For example, if `billions_of_numbers` is an array of numbers, the expression

    .18 * billions_of_numbers

gives a new array of numbers that contains the result of multiplying **each number** in `billions_of_numbers` by .18.  Arrays are not limited to numbers; we can also put all the words in a book into an array of strings.

Concretely, an array is a **sequence of values of the same type**. 

## 2.1. Making arrays

First, let's learn how to manually input values into an array. This typically isn't how programs work. Normally, we create arrays by loading them from an external source, like a data file.

To create an array by hand, call the function `make_array`.  Each argument you pass to `make_array` will be in the array it returns.  Run this cell to see an example:

In [19]:
make_array(0.125, 4.75, -1.3)

array([ 0.125,  4.75 , -1.3  ])

Each value in an array (in the above case, the numbers 0.125, 4.75, and -1.3) is called an *element* of that array.

Arrays themselves are also values, just like numbers and strings.  That means you can assign them to names or use them as arguments to functions. For example, `len(<some_array>)` returns the number of elements in `some_array`.

**Question 2.1.1.** Make an array containing the numbers 0, 1, -1, and $\pi$, in that order.  Name it `interesting_numbers`.  

*Hint:* How did you get the value $\pi$ in an earlier lab?  You can refer to it the same way here. The `math` module has been imported at the top of this notebook.

<!--
BEGIN QUESTION
name: q211
-->

In [70]:

interesting_numbers = make_array(0, 1, -1, math.pi)
interesting_numbers

array([ 0.        ,  1.        , -1.        ,  3.14159265])

In [71]:
grader.check("q211")

q211 results: All test cases passed!

**Question 2.1.2.** Make an array containing the five strings `"Hello"`, `","`, `" "`, `"world"`, and `"!"`.  (The third one is a single space inside quotes.)  Name it `hello_world_components`.

*Note:* If you evaluate `hello_world_components`, you'll notice some extra information in addition to its contents: `dtype='<U5'`.  That's just NumPy's extremely cryptic way of saying that the data types in the array are strings.

<!--
BEGIN QUESTION
name: q212
-->

In [22]:
import numpy as np

# Create the NumPy array with the specified strings
hello_world_components = np.array(["Hello", ",", " ", "world", "!"])
hello_world_components


array(['Hello', ',', ' ', 'world', '!'],
      dtype='<U5')

In [24]:
grader.check("q212")

q212 results: All test cases passed!

###  `np.arange`
Arrays are provided by a package called [NumPy](http://www.numpy.org/) (pronounced "NUM-pie"). The package is called `numpy`, but it's standard to rename it `np` for brevity.  You can do that with:

    import numpy as np

Very often in data science, we want to work with many numbers that are evenly spaced within some range.  NumPy provides a special function for this called `arange`.  The line of code `np.arange(start, stop, step)` evaluates to an array with all the numbers starting at `start` and counting up by `step`, stopping **before** `stop` is reached.

Run the following cells to see some examples!

In [25]:
# This array starts at 1 and counts up by 2
# and then stops before 6
np.arange(1, 6, 2)

array([1, 3, 5])

In [26]:
# This array doesn't contain 9
# because np.arange stops *before* the stop value is reached
np.arange(4, 9, 1)

array([4, 5, 6, 7, 8])

**Question 2.1.3.** Import `numpy` as `np` and then use `np.arange` to create an array with the multiples of 99 from 0 up to (**and including**) 9999.  (So its elements are 0, 99, 198, 297, etc.)

<!--
BEGIN QUESTION
name: q213
-->

In [27]:
import numpy as np

# Create an array with multiples of 99 from 0 to 9999
multiples_of_99 = np.arange(0, 10000, 99)
multiples_of_99


array([   0,   99,  198,  297,  396,  495,  594,  693,  792,  891,  990,
       1089, 1188, 1287, 1386, 1485, 1584, 1683, 1782, 1881, 1980, 2079,
       2178, 2277, 2376, 2475, 2574, 2673, 2772, 2871, 2970, 3069, 3168,
       3267, 3366, 3465, 3564, 3663, 3762, 3861, 3960, 4059, 4158, 4257,
       4356, 4455, 4554, 4653, 4752, 4851, 4950, 5049, 5148, 5247, 5346,
       5445, 5544, 5643, 5742, 5841, 5940, 6039, 6138, 6237, 6336, 6435,
       6534, 6633, 6732, 6831, 6930, 7029, 7128, 7227, 7326, 7425, 7524,
       7623, 7722, 7821, 7920, 8019, 8118, 8217, 8316, 8415, 8514, 8613,
       8712, 8811, 8910, 9009, 9108, 9207, 9306, 9405, 9504, 9603, 9702,
       9801, 9900, 9999])

In [28]:
grader.check("q213")

q213 results: All test cases passed!

## 2.2. Working with single elements of arrays ("indexing")
Let's work with a more interesting dataset.  The next cell creates an array called `population_amounts` that includes estimated world populations of every year from **1950** to roughly the present.  (The estimates come from the US Census Bureau website.)

Rather than type in the data manually, we've loaded them from a file on your computer called `world_population.csv`.  You'll learn how to do that later in this lab!

In [29]:
population_amounts = Table.read_table("world_population.csv").column("Population")
population_amounts

array([2557628654, 2594939877, 2636772306, 2682053389, 2730228104,
       2782098943, 2835299673, 2891349717, 2948137248, 3000716593,
       3043001508, 3083966929, 3140093217, 3209827882, 3281201306,
       3350425793, 3420677923, 3490333715, 3562313822, 3637159050,
       3712697742, 3790326948, 3866568653, 3942096442, 4016608813,
       4089083233, 4160185010, 4232084578, 4304105753, 4379013942,
       4451362735, 4534410125, 4614566561, 4695736743, 4774569391,
       4856462699, 4940571232, 5027200492, 5114557167, 5201440110,
       5288955934, 5371585922, 5456136278, 5538268316, 5618682132,
       5699202985, 5779440593, 5857972543, 5935213248, 6012074922,
       6088571383, 6165219247, 6242016348, 6318590956, 6395699509,
       6473044732, 6551263534, 6629913759, 6709049780, 6788214394,
       6866332358, 6944055583, 7022349283, 7101027895, 7178722893,
       7256490011])

Here's how we get the first element of `population_amounts`, which is the world population in the first year in the dataset, 1950.

In [30]:
population_amounts.item(0)

2557628654

The value of that expression is the number 2,557,628,654 (around 2.5 billion), because that's the first thing in the array `population_amounts`.

Notice that we wrote `.item(0)`, not `.item(1)`, to get the first element.  This is a weird, but ultimately very useful, convention in computer science.  **0 is called the *index* of the first item.**  It's the number of elements that appear *before* that item.  So 3 is the index of the 4th item.

You can also think of the index as the number of jumps to get to that element. Since you enter at the beginning of the array, getting to the first element requires no jumps. The second element requires 1 jump, and so on.

Here are some more examples.  In the examples, we've given names to the things we get out of `population_amounts`.  Read and run each cell.

In [31]:
# The 13th element in the array is the population
# in 1962 (which is 1950 + 12).
population_1962 = population_amounts.item(12)
population_1962

3140093217

In [32]:
# The 66th element is the population in 2015.
population_2015 = population_amounts.item(65)
population_2015

7256490011

In [33]:
# The array has only 66 elements, so this doesn't work.
# (There's no element with 66 other elements before it.)
population_2016 = population_amounts.item(66)
population_2016

IndexError: index 66 is out of bounds for axis 0 with size 66

(Note that this error message is a bit cryptic. But it does point you at the source of the error and 'out of bounds' is suggestive of the problem.  Don't worry about the extra details that you don't understand. Usually just knowing there is an error is enough.)

Since `make_array` returns an array, we can call `.item(3)` on its output to get its 4th element, just like we "chained" together calls to the method `replace` earlier.

In [34]:
make_array(-1, -3, 4, -2).item(3)

-2

**Question 2.2.1.** Set `population_1973` to the world population in 1973, by getting the appropriate element from `population_amounts` using `item`.

<!--
BEGIN QUESTION
name: q221
-->

In [35]:
population_1973 = population_amounts.item(23)
population_1973

3942096442

In [36]:
grader.check("q221")

q221 results: All test cases passed!

## 2.3. Doing something to every element of an array
Arrays are primarily useful for doing the same operation many times, so we don't often have to use `.item` and work with single elements.

##### Rounding
Here is one simple question we might ask about world population:

> How big was the population in each year, rounded to the nearest million?

Rounding is often used with large numbers when we don't need as much precision in our numbers. One example of this is when we present data in tables and visualizations. 

We could try to answer our question using the `round` function that is built into Python and the `item` method you just saw. 

**Note:** the `round` function takes in two arguments: the number to be rounded, and the number of decimal places to round to. The second argument can be thought of as how many steps right or left you move from the decimal point. Negative numbers tell us to move left, and positive numbers tell us to move right. So, if we have `round(1234.5, -2)`, it means that we should move two places left, and then make all numbers to the right of this place zeroes. This would output the number 1200.0. On the other hand, if we have `round(6.789, 1)`, we should move one place right, and then make all numbers to the right of this place zeroes. This would output the number 6.8.

In [37]:
population_1950_magnitude = round(population_amounts.item(0), -6)
population_1951_magnitude = round(population_amounts.item(1), -6)
population_1952_magnitude = round(population_amounts.item(2), -6)
population_1953_magnitude = round(population_amounts.item(3), -6)

But this is tedious and doesn't really take advantage of the fact that we are using a computer.

Instead, NumPy provides its own version of `round` that rounds each element of an array.  It takes in two arguments: a single array of numbers, and the number of decimal places to round to.  It returns an array of the same length, where the first element of the result is the first element of the argument rounded, and so on.

**Question 2.3.1.** Use `np.round` to compute the world population in every year, rounded to the nearest million (6 zeroes).  Give the result (an array of 66 numbers) the name `population_rounded`.  Your code should be very short.

<!--
BEGIN QUESTION
name: q231
-->

In [38]:
population_rounded = np.round(population_amounts / 1_000_000) * 1_000_000
population_rounded

array([  2.55800000e+09,   2.59500000e+09,   2.63700000e+09,
         2.68200000e+09,   2.73000000e+09,   2.78200000e+09,
         2.83500000e+09,   2.89100000e+09,   2.94800000e+09,
         3.00100000e+09,   3.04300000e+09,   3.08400000e+09,
         3.14000000e+09,   3.21000000e+09,   3.28100000e+09,
         3.35000000e+09,   3.42100000e+09,   3.49000000e+09,
         3.56200000e+09,   3.63700000e+09,   3.71300000e+09,
         3.79000000e+09,   3.86700000e+09,   3.94200000e+09,
         4.01700000e+09,   4.08900000e+09,   4.16000000e+09,
         4.23200000e+09,   4.30400000e+09,   4.37900000e+09,
         4.45100000e+09,   4.53400000e+09,   4.61500000e+09,
         4.69600000e+09,   4.77500000e+09,   4.85600000e+09,
         4.94100000e+09,   5.02700000e+09,   5.11500000e+09,
         5.20100000e+09,   5.28900000e+09,   5.37200000e+09,
         5.45600000e+09,   5.53800000e+09,   5.61900000e+09,
         5.69900000e+09,   5.77900000e+09,   5.85800000e+09,
         5.93500000e+09,

In [39]:
grader.check("q231")

q231 results: All test cases passed!

What you just did is called **elementwise** application of `np.round`, since `np.round` operates separately on each element of the array that it's called on. Here's a picture of what's going on:

<img src="array_round.jpg">

The textbook's [section](https://www.inferentialthinking.com/chapters/05/1/Arrays)  on arrays has a useful list of NumPy functions that are designed to work elementwise, like `np.round`.

##### Arithmetic
Arithmetic also works elementwise on arrays, meaning that if you perform an arithmetic operation (like subtraction, division, etc) on an array, Python will do the operation to every element of the array individually and return an array of all of the results. For example, you can divide all the population numbers by 1 billion to get numbers in billions:

In [40]:
population_in_billions = population_amounts / 1000000000
population_in_billions

array([ 2.55762865,  2.59493988,  2.63677231,  2.68205339,  2.7302281 ,
        2.78209894,  2.83529967,  2.89134972,  2.94813725,  3.00071659,
        3.04300151,  3.08396693,  3.14009322,  3.20982788,  3.28120131,
        3.35042579,  3.42067792,  3.49033371,  3.56231382,  3.63715905,
        3.71269774,  3.79032695,  3.86656865,  3.94209644,  4.01660881,
        4.08908323,  4.16018501,  4.23208458,  4.30410575,  4.37901394,
        4.45136274,  4.53441012,  4.61456656,  4.69573674,  4.77456939,
        4.8564627 ,  4.94057123,  5.02720049,  5.11455717,  5.20144011,
        5.28895593,  5.37158592,  5.45613628,  5.53826832,  5.61868213,
        5.69920299,  5.77944059,  5.85797254,  5.93521325,  6.01207492,
        6.08857138,  6.16521925,  6.24201635,  6.31859096,  6.39569951,
        6.47304473,  6.55126353,  6.62991376,  6.70904978,  6.78821439,
        6.86633236,  6.94405558,  7.02234928,  7.10102789,  7.17872289,
        7.25649001])

You can do the same with addition, subtraction, multiplication, and exponentiation (`**`).

## 3. Creating Tables

An array is useful for describing a single attribute of each element in a collection. For example, let's say our collection is all US States. Then an array could describe the land area of each state. 

Tables extend this idea by containing multiple columns stored and represented as arrays, each one describing a different attribute for every element of a collection. In this way, tables allow us to not only store data about many entities but to also contain several kinds of data about each entity. 

For example, in the cell below we have two arrays. The first one, `population_amounts`, was defined above in section 2.2 and contains the world population in each year (estimated by the US Census Bureau). The second array, `years`, contains the years themselves. These elements are in order, so the year and the world population for that year have the same index in their corresponding arrays.

In [41]:
# Just run this cell

years = np.arange(1950, 2015+1)
print("Population column:", population_amounts)
print("Years column:", years)

Population column: [2557628654 2594939877 2636772306 2682053389 2730228104 2782098943
 2835299673 2891349717 2948137248 3000716593 3043001508 3083966929
 3140093217 3209827882 3281201306 3350425793 3420677923 3490333715
 3562313822 3637159050 3712697742 3790326948 3866568653 3942096442
 4016608813 4089083233 4160185010 4232084578 4304105753 4379013942
 4451362735 4534410125 4614566561 4695736743 4774569391 4856462699
 4940571232 5027200492 5114557167 5201440110 5288955934 5371585922
 5456136278 5538268316 5618682132 5699202985 5779440593 5857972543
 5935213248 6012074922 6088571383 6165219247 6242016348 6318590956
 6395699509 6473044732 6551263534 6629913759 6709049780 6788214394
 6866332358 6944055583 7022349283 7101027895 7178722893 7256490011]
Years column: [1950 1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962 1963 1964
 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 1977 1978 1979
 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992 1993 1994
 

Suppose we want to answer this question:

> In which year did the world's population cross 6 billion?

You could technically answer this question just from staring at the arrays, but it's a bit convoluted, since you would have to count the position where the population first crossed 6 billion, then find the corresponding element in the years array. In cases like these, it might be easier to put the data into a *`Table`*, a 2-dimensional type of dataset. 

The expression below:

- creates an empty table using the expression `Table()`,
- adds two columns by calling `with_columns` with four arguments,
- assigns the result to the name `population`, and finally
- evaluates `population` so that we can see the table.

The strings `"Year"` and `"Population"` are column labels that we have chosen. The names `population_amounts` and `years` were assigned above to two arrays of the **same length**. The function `with_columns` (you can find the documentation in the [Python Reference](http://data8.org/sp22/python-reference.html)) takes in alternating strings (to represent column labels) and arrays (representing the data in those columns). The strings and arrays are separated by commas.

In [42]:
population = Table().with_columns(
    "Population", population_amounts,
    "Year", years
)
population

Population | Year
2557628654 | 1950
2594939877 | 1951
2636772306 | 1952
2682053389 | 1953
2730228104 | 1954
2782098943 | 1955
2835299673 | 1956
2891349717 | 1957
2948137248 | 1958
3000716593 | 1959
... (56 rows omitted)

Now the data is combined into a single table! It's much easier to parse this data. If you need to know what the population was in 1959, for example, you can tell from a single glance.

**Question 3.1.** In the cell below, we've created 2 arrays. Using the steps above, assign `top_10_movies` to a table that has two columns called "Name" and "Rating", which hold `top_10_movie_names` and `top_10_movie_ratings` respectively.

<!--
BEGIN QUESTION
name: q31
-->

In [43]:
top_10_movie_names = make_array(
        'The Shawshank Redemption (1994)',
        'The Godfather (1972)',
        'The Godfather: Part II (1974)',
        'Pulp Fiction (1994)',
        "Schindler's List (1993)",
        'The Lord of the Rings: The Return of the King (2003)',
        '12 Angry Men (1957)',
        'The Dark Knight (2008)',
        'Il buono, il brutto, il cattivo (1966)',
        'The Lord of the Rings: The Fellowship of the Ring (2001)')
top_10_movie_ratings = make_array(9.2, 9.2, 9., 8.9, 8.9, 8.9, 8.9, 8.9, 8.9, 8.8)

top_10_movies = Table().with_columns(
    "Name", top_10_movie_names,
    "Rating", top_10_movie_ratings
)

# We've put this next line here 
# so your table will get printed out 
# when you run this cell.
top_10_movies

Name                                                     | Rating
The Shawshank Redemption (1994)                          | 9.2
The Godfather (1972)                                     | 9.2
The Godfather: Part II (1974)                            | 9
Pulp Fiction (1994)                                      | 8.9
Schindler's List (1993)                                  | 8.9
The Lord of the Rings: The Return of the King (2003)     | 8.9
12 Angry Men (1957)                                      | 8.9
The Dark Knight (2008)                                   | 8.9
Il buono, il brutto, il cattivo (1966)                   | 8.9
The Lord of the Rings: The Fellowship of the Ring (2001) | 8.8

In [44]:
grader.check("q31")

q31 results: All test cases passed!

#### Loading a table from a file

In most cases, we aren't going to go through the trouble of typing in all the data manually. Instead, we load them in from an external source, like a data file. There are many formats for data files, but CSV ("comma-separated values") is the most common.

`Table.read_table(...)` takes one argument (a path to a data file in **string** format) and returns a table.  

**Question 3.2.** `imdb.csv` contains a table of information about the 250 highest-rated movies on IMDb.  Load it as a table called `imdb`.

(You may remember working with this table in a previous lab!)

<!--
BEGIN QUESTION
name: q32
-->

In [45]:
imdb = Table.read_table("imdb.csv")
imdb

Votes  | Rating | Title                 | Year | Decade
88355  | 8.4    | M                     | 1931 | 1930
132823 | 8.3    | Singin' in the Rain   | 1952 | 1950
74178  | 8.3    | All About Eve         | 1950 | 1950
635139 | 8.6    | Léon                  | 1994 | 1990
145514 | 8.2    | The Elephant Man      | 1980 | 1980
425461 | 8.3    | Full Metal Jacket     | 1987 | 1980
441174 | 8.1    | Gone Girl             | 2014 | 2010
850601 | 8.3    | Batman Begins         | 2005 | 2000
37664  | 8.2    | Judgment at Nuremberg | 1961 | 1960
46987  | 8      | Relatos salvajes      | 2014 | 2010
... (240 rows omitted)

In [46]:
grader.check("q32")

q32 results: All test cases passed!

Where did `imdb.csv` come from? Remember that you unzipped a file to find the notebook `lab04_104.ipynb` for this lab.  Some other files were generated in the same directory as the notebook, including a file called `imdb.csv`.  Go ahead and take a look at this file.
(If you open a Unix terminal and move to the directory containing this file, the command `more imdb.csv` will let you look at the contents.)
The `.csv` filename ending says that this file is in the [CSV (comma-separated value) format](http://edoceo.com/utilitas/csv-file-format).

## 4. More Table Operations!

Now that you've worked with arrays, let's add a few more methods to the list of table operations that you saw in a previous lab.

### `column`

`column` takes the column name of a table (in string format) as its argument and returns the values in that column as an **array**. 

In [47]:
# Returns an array of movie names
top_10_movies.column('Name')

array(['The Shawshank Redemption (1994)', 'The Godfather (1972)',
       'The Godfather: Part II (1974)', 'Pulp Fiction (1994)',
       "Schindler's List (1993)",
       'The Lord of the Rings: The Return of the King (2003)',
       '12 Angry Men (1957)', 'The Dark Knight (2008)',
       'Il buono, il brutto, il cattivo (1966)',
       'The Lord of the Rings: The Fellowship of the Ring (2001)'],
      dtype='<U56')

### `take`
The table method `take` takes as its argument an array of numbers.  Each number should be the index of a row in the table.  It returns a **new table** with only those rows. 

You'll usually want to use `take` in conjunction with `np.arange` to take the first few rows of a table.

In [48]:
# Take first 5 movies of top_10_movies
top_10_movies.take(np.arange(0, 5, 1))

Name                            | Rating
The Shawshank Redemption (1994) | 9.2
The Godfather (1972)            | 9.2
The Godfather: Part II (1974)   | 9
Pulp Fiction (1994)             | 8.9
Schindler's List (1993)         | 8.9

The next three questions will give you practice with combining the operations you've learned in this lab and the previous one to answer questions about the `population` and `imdb` tables. First, check out the `population` table from section 2.

In [49]:
# Run this cell to display the population table.
# Display the first 20 rows of the population table
population.take(np.arange(10))


Population | Year
2557628654 | 1950
2594939877 | 1951
2636772306 | 1952
2682053389 | 1953
2730228104 | 1954
2782098943 | 1955
2835299673 | 1956
2891349717 | 1957
2948137248 | 1958
3000716593 | 1959

**Question 4.1.** Find the year when the world population first went above 6 billion. Assign the year to `year_population_crossed_6_billion`.

*Approach 1*: In the previous cell, use `take` to explore the table, looking for the row.
            The problem is that you need to search for the correct rows.
            The next approach does the searching for you.

*Approach 2*: In the previous cell, use a table method from the previous lab to build a subtable of years with population greater than 6 billion. Find the year easily in this table.

<!--
BEGIN QUESTION
name: q41
-->

In [72]:
subtable = 
year_population_crossed_6_billion = ...
year_population_crossed_6_billion



Ellipsis

In [73]:
grader.check("q41")

q41 results:
    q41 - 1 result:
        ❌ Test case failed
        Trying:
            import numpy as np
        Expecting nothing
        ok
        Trying:
            type(year_population_crossed_6_billion) == int or type(year_population_crossed_6_billion) == np.int64
        Expecting:
            True
        **********************************************************************
        Line 3, in q41 0
        Failed example:
            type(year_population_crossed_6_billion) == int or type(year_population_crossed_6_billion) == np.int64
        Expected:
            True
        Got:
            False

    q41 - 2 result:
        ❌ Test case failed
        Trying:
            year_population_crossed_6_billion == 1999
        Expecting:
            True
        **********************************************************************
        Line 1, in q41 1
        Failed example:
            year_population_crossed_6_billion == 1999
        Expected:
            True
        Got:
            False

**Question 4.2.** Find the average rating for movies released before the year 2000 and the average rating for movies released in the year 2000 or after for the movies in `imdb`.

*Hint*: Think of the steps you need to do (take the average, find the ratings, find movies released in 20th/21st centuries), and try to put them in an order that makes sense.

<!--
BEGIN QUESTION
name: q42
-->

In [89]:
imdb

Votes  | Rating | Title                 | Year | Decade
88355  | 8.4    | M                     | 1931 | 1930
132823 | 8.3    | Singin' in the Rain   | 1952 | 1950
74178  | 8.3    | All About Eve         | 1950 | 1950
635139 | 8.6    | Léon                  | 1994 | 1990
145514 | 8.2    | The Elephant Man      | 1980 | 1980
425461 | 8.3    | Full Metal Jacket     | 1987 | 1980
441174 | 8.1    | Gone Girl             | 2014 | 2010
850601 | 8.3    | Batman Begins         | 2005 | 2000
37664  | 8.2    | Judgment at Nuremberg | 1961 | 1960
46987  | 8      | Relatos salvajes      | 2014 | 2010
... (240 rows omitted)

In [87]:

before_2000 = np.average(imdb.where('Year', are.below(2000)).column('Rating'))
after_or_in_2000 = np.average(imdb.where('Year', are.above_or_equal_to(2000)).column('Rating'))



print("Average before 2000 rating:", before_2000)
print("Average after or in 2000 rating:", after_or_in_2000)


Average before 2000 rating: 8.2783625731
Average after or in 2000 rating: 8.23797468354


In [88]:
grader.check("q42")

Exception: Could not create the log file as the file system is read-only. Please contact your instructor before continuing on this assignment.

**Question 4.3.** Here's a challenge: Find the number of movies that came out in *even* years.

*Hint:* The operator `%` computes the remainder when dividing by a number.  So `5 % 2` is 1 and `6 % 2` is 0.  A number is even if the remainder is 0 when you divide by 2.

*Hint 2:* `%` can be used on arrays, operating elementwise like `+` or `*`.  So `make_array(5, 6, 7) % 2` is `array([1, 0, 1])`.

*Hint 3:* Add a new column called "Year Remainder" that's the remainder when each movie's release year is divided by 2. To do this, you can use `tbl.with_column(col_name, col_values)`: a table method that takes in the name of the new column and an array of values and returns a copy of the original `tbl` with the new column.  Then, use `where` to find rows where that new column is equal to 0.  Finally, use `num_rows` to count the number of such rows.

*Note:* These steps can be chained in one single statement, or broken up across several lines with intermediate names assigned. You’re always welcome to break down problems however you wish!

<!--
BEGIN QUESTION
name: q43
-->

In [52]:
# Add a new column for the year remainder
imdb_with_remainder = imdb.with_column('Year Remainder', imdb.column('Year') % 2)

# Count the number of movies with even years
num_even_year_movies = imdb_with_remainder.where('Year Remainder', 0).num_rows


In [53]:
grader.check("q43")

q43 results: All test cases passed!

## 5. More Array Practice

The following questions will be great practice for learning how to generate arrays in different ways, as well as executing different methods of array arithmetic!

###  5.1 `np.arange` (cont.)

##### Temperature readings
NOAA (the US National Oceanic and Atmospheric Administration) operates weather stations that measure surface temperatures at different sites around the United States.  The hourly readings are [publicly available](http://www.ncdc.noaa.gov/qclcd/QCLCD?prior=N).

Suppose we download all the hourly data from the Birmingham, Alabama site for the month of December 2021.  To analyze the data, we want to know when each reading was taken, but we find that the data don't include the timestamps of the readings (the time at which each one was taken).

However, we know the first reading was taken at the first instant of December 2021 (midnight on December 1st) and each subsequent reading was taken exactly 1 hour after the last.

**Question 5.1.1.** Create an array of the *time, in seconds, since the start of the month* at which **each hourly reading** was taken.  Name it `collection_times`.

*Hint 1:* There were 31 days in December, which is equivalent to ($31 \times 24$) hours or ($31 \times 24 \times 60 \times 60$) seconds.  So your array should have $31 \times 24$ elements in it.

*Hint 2:* The `len` function works on arrays, too!  If your `collection_times` isn't passing the tests, check its length and make sure it has $31 \times 24$ elements.

<!--
BEGIN QUESTION
name: q511
-->

In [54]:
# Total number of hours in December 2021
total_hours = 31 * 24

# Create an array of collection times in seconds
collection_times = np.arange(total_hours) * 3600  # 3600 seconds in an hour


In [55]:
grader.check("q511")

q511 results: All test cases passed!

### 5.2 Doing something to every element of an array (cont.)

Calculate a tip on several restaurant bills at once (in this case just 3):

In [56]:
restaurant_bills = make_array(20.12, 39.90, 31.01)
print("Restaurant bills:\t", restaurant_bills)

# Array multiplication
tips = .2 * restaurant_bills
print("Tips:\t\t\t", tips)

Restaurant bills:	 [ 20.12  39.9   31.01]
Tips:			 [ 4.024  7.98   6.202]


<img src="array_multiplication.jpg">

**Question 5.2.1** Suppose the total charge at a restaurant is the original bill plus the tip. If the tip is 20%, that means we can multiply the original bill by 1.2 to get the total charge.  Compute the total charge for each bill in `restaurant_bills`, and assign the resulting array to `total_charges`.

<!--
BEGIN QUESTION
name: q521
-->

In [57]:
total_charges = restaurant_bills * 1.2
total_charges

array([ 24.144,  47.88 ,  37.212])

In [58]:
grader.check("q521")

q521 results: All test cases passed!

**Question 5.2.2.** The array `more_restaurant_bills` contains 100,000 bills!  Compute the total charge for each one in `more_total_charges`.  How is your code different?

<!--
BEGIN QUESTION
name: q522
-->

In [79]:



more_restaurant_bills = Table.read_table("more_restaurant_bills.csv").column("Bill")
more_total_charges = more_restaurant_bills * 1.2
more_total_charges




array([ 20.244,  20.892,  12.216, ...,  19.308,  18.336,  35.664])

In [80]:
grader.check("q522")

q522 results: All test cases passed!

The function `sum` takes a single array of numbers as its argument.  It returns the sum of all the numbers in that array (so it returns a single number, not an array).

**Question 5.2.3.** What was the sum of all the bills in `more_restaurant_bills`, *including tips*?

<!--
BEGIN QUESTION
name: q523
-->

array([ 20.12,  39.9 ,  31.01])

In [96]:



sum_of_bills = ...

sum_of_bills

AttributeError: 'list' object has no attribute 'sum'

In [86]:
grader.check("q523")

q523 results:
    q523 - 1 result:
        ❌ Test case failed
        Trying:
            round(sum_of_bills, 2) == 1795730.06
        Expecting:
            True
        **********************************************************************
        Line 1, in q523 0
        Failed example:
            round(sum_of_bills, 2) == 1795730.06
        Expected:
            True
        Got:
            False

**Question 5.2.4.** The powers of 2 ($2^0 = 1$, $2^1 = 2$, $2^2 = 4$, etc) arise frequently in computer science.  (For example, you may have noticed that storage on smartphones or USBs come in powers of 2, like 16 GB, 32 GB, or 64 GB.)  Use `np.arange` and the exponentiation operator `**` to compute the first 30 powers of 2, starting from `2^0`.

*Hint 1:* `np.arange(1, 2**30, 1)` creates an array with $2^{30}$ elements and **will crash your kernel**.

*Hint 2:* Part of your solution will involve `np.arange`, but your array shouldn't have more than 30 elements.

<!--
BEGIN QUESTION
name: q524
-->

In [67]:
exponents = np.arange(30)

powers_of_2 = 2 ** exponents
powers_of_2

array([        1,         2,         4,         8,        16,        32,
              64,       128,       256,       512,      1024,      2048,
            4096,      8192,     16384,     32768,     65536,    131072,
          262144,    524288,   1048576,   2097152,   4194304,   8388608,
        16777216,  33554432,  67108864, 134217728, 268435456, 536870912])

In [68]:
grader.check("q524")

q524 results: All test cases passed!

---

Congratulations on finishing lab!  Have a piece of cake!

<img src="thiebaud_cakes.jpg" alt="drawing" width="300"/>

[Wayne Thiebaud, Cakes, 1963](https://www.nga.gov/collection/art-object-page.72040.html)


---

To double-check your work, the cell below will rerun all of the autograder tests.

In [ ]:
grader.check_all()

## Submission

**Important submission information:** 

- Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output. 
- Be sure to run the tests and verify that they all pass.
- Then choose **Save** from the **File** menu to save your file.
- Then, go to Canvas and submit the file as your solution. (Only the prescribed team member needs to submit.)